### DISCUSSION TITLE
Is there a way to configure the version of rust-analyzer used in zed?

### DISCUSSION THREAD

**Discussion by emmaling27** [zed-industries/zed#12155] [Feature Requests]
rust-analyzer tends to break unexpectedly on upgrades. I noticed Zed automatically upgrades rust-analyzer, but this makes zed unusable for me (the 2024-05-20 release crashes in my codebase). Is there a way to pin the version of rust-analyzer that zed uses?

**Comment by SomeoneToIgnore** [comment]
IIRC right now, there's no good way to tell Zed which version of the language server to use, but there's a way to use a specific rust-analyzer binary that will not be [re]downloaded or anyhow managed by Zed: https://github.com/zed-industries/zed/pull/9293

**Reply by emmaling27** [reply]
Thanks! This worked for me. It would be nice if this was a feature in the future, something like 
```
{
  "lsp": {
    "rust-analyzer": {
      "version": "<a-stable-version>"
    }
  }
}
```
Any idea how much work that would be or a good entrypoint if I were to take a stab at it?

**Reply by emmaling27** [reply]
Actually, now that I'm double checking, this doesn't solve my problem. rust-analyzer still crashes on the "stable version" I chose (which works fine in VS code). Not sure what else to do to debug. Language server logs didn't show anything related to rust-analyzer but zed logs show repeated 
```
crates/project/src/project.rs:7020: server shut down
2024-05-24T21:25:04-07:00 [WARN] Generic lsp request to rust-analyzer failed: server shut down
2024-05-24T21:25:04-07:00 [ERROR] crates/editor/src/element.rs:287: server shut down
```

**Reply by SomeoneToIgnore** [reply]
> Any idea how much work that would be or a good entrypoint if I were to take a stab at it?

I do not think it's very hard, check out https://github.com/zed-industries/zed/pull/9293 that added a path properly, and consider vue retrieval code, that harcodes its current version: https://github.com/zed-industries/zed/blob/d5fe2c85d80288a0f50eb1c2dfb1e82f11046402/extensions/vue/src/vue.rs#L29

Some fusion of both places has to be done, so the user can set something in its settings, and Zed gets that version number during the LSP server fetch.

Seems that the biggest complexity would be to ensure that there are some pop-up notifications during fetch errors, and those are not too noisy.

> 2024-05-24T21:25:04-07:00 [WARN] Generic lsp request to rust-analyzer failed: server shut down

That looks like r-a is panicking and restarting in a loop.
There's a slight chance something is wrong in Zed LSP init code, but since it used to work before, not likely?
I would check out https://rust-analyzer.github.io/manual.html#troubleshooting first and try to send r-a logs into a file (if it really panics in a loop, there will be a lot of logs to process).
